### PR TITLE
[4/4] Port spec tests for limitToLast

### PR DESCRIPTION
**Pull request by wu-hui**

### CODE REVIEW

File: Firestore/Example/Tests/SpecTests/FSTSyncEngineTestDriver.h [hunk 1]
```diff
@@ -48,6 +50,11 @@ class Persistence;
 namespace core = firebase::firestore::core;
 namespace local = firebase::firestore::local;
 namespace model = firebase::firestore::model;
+namespace nanopb = firebase::firestore::nanopb;
+
+using ActiveTargetMap =
```
Comment (wilhuff) [inline Firestore/Example/Tests/SpecTests/FSTSyncEngineTestDriver.h]: Could you add comments to this declaration indicating what this should contain? Previously this was just a map of target id to query data, but now I have no idea what's going on here.

File: Firestore/Example/Tests/SpecTests/FSTSyncEngineTestDriver.h [hunk 2]
```diff
@@ -31,6 +32,7 @@
 #include "Firestore/core/src/firebase/firestore/model/mutation.h"
 #include "Firestore/core/src/firebase/firestore/model/snapshot_version.h"
 #include "Firestore/core/src/firebase/firestore/model/types.h"
+#include "Firestore/core/src/firebase/firestore/nanopb/nanopb_util.h"
```
Comment (wilhuff) [inline Firestore/Example/Tests/SpecTests/FSTSyncEngineTestDriver.h]: Include nanopb/byte_string.h to get a declaration of `ByteString`.

File: Firestore/Example/Tests/SpecTests/FSTSyncEngineTestDriver.mm [hunk 3]
```diff
@@ -473,11 +475,11 @@ - (void)receiveWatchStreamError:(int)errorCode userInfo:(NSDictionary<NSString *
   return _datastore->ActiveTargets();
 }
 
-- (const std::unordered_map<TargetId, QueryData> &)expectedActiveTargets {
+- (const ActiveTargetMap &)expectedActiveTargets {
   return _expectedActiveTargets;
 }
 
-- (void)setExpectedActiveTargets:(const std::unordered_map<TargetId, QueryData> &)targets {
+- (void)setExpectedActiveTargets:(const ActiveTargetMap &)targets {
```
Comment (wilhuff) [inline Firestore/Example/Tests/SpecTests/FSTSyncEngineTestDriver.mm]: You could pass by value and `std::move(targets)` to potentially avoid an O(n) copy of the target map. (The caller would still have to `std::move` to save the copy).

File: Firestore/core/src/firebase/firestore/core/sync_engine.cc [hunk 4]
```diff
@@ -110,10 +110,23 @@ ViewSnapshot SyncEngine::InitializeViewAndComputeSnapshot(const Query& query,
   DocumentMap docs = local_store_->ExecuteQuery(query);
   DocumentKeySet remote_keys = local_store_->GetRemoteDocumentKeys(target_id);
 
+  // If there are already queries mapped to the target id, create a synthesized
+  // target change to apply the sync state from those queries to the new query.
+  auto current_sync_state = SyncState::None;
+  TargetChange synthesized_current_change;
```
Comment (wilhuff) [inline Firestore/core/src/firebase/firestore/core/sync_engine.cc]: This isn't equivalent to the android version. There this defaults to null, and only if there's a synthesized change does it get set to a non-null value.

Here you're creating a default TargetChange. This type should likely be `absl::optional<TargetChange>` (which is the type of the parameter to `ViewChange::ApplyChanges`.

File: Firestore/core/src/firebase/firestore/remote/remote_event.h [hunk 5]
```diff
@@ -74,6 +74,9 @@ class TargetChange {
  public:
   TargetChange() = default;
 
+  explicit TargetChange(bool current) : current_{current} {
```
Comment (wilhuff) [inline Firestore/core/src/firebase/firestore/remote/remote_event.h]: The Android version has this as a static factory named `createSynthesizedTargetChangeForCurrentChange`. This is considerably more explicit and self-documenting than just a constructor taking a boolean. Please port.

File: Firestore/Example/Tests/SpecTests/FSTSyncEngineTestDriver.h (hunk 2 above)
Comment (wu-hui) [reply Firestore/Example/Tests/SpecTests/FSTSyncEngineTestDriver.h]: Done.

File: Firestore/Example/Tests/SpecTests/FSTSyncEngineTestDriver.h (hunk 1 above)
Comment (wu-hui) [reply Firestore/Example/Tests/SpecTests/FSTSyncEngineTestDriver.h]: Added comment.

File: Firestore/core/src/firebase/firestore/core/sync_engine.cc (hunk 4 above)
Comment (wu-hui) [reply Firestore/core/src/firebase/firestore/core/sync_engine.cc]: Done.

File: Firestore/core/src/firebase/firestore/remote/remote_event.h (hunk 5 above)
Comment (wu-hui) [reply Firestore/core/src/firebase/firestore/remote/remote_event.h]: Done.

File: Firestore/Example/Tests/SpecTests/FSTSyncEngineTestDriver.mm (hunk 3 above)
Comment (wu-hui) [reply Firestore/Example/Tests/SpecTests/FSTSyncEngineTestDriver.mm]: Done.

**Pull request merged**